libgobuster/dir.go

```diff
@@ -59,6 +61,22 @@ func MakeRequest(s *State, fullUrl, cookie string) (*int, *int64) {
 		req.Header.Set("User-Agent", s.UserAgent)
 	}
 
+	if s.RandomUserAgent && s.UserAgent == "" {
+		b, err := ioutil.ReadFile(s.UserAgentsFile)
```

Comment:
We could probably make this more efficient by reading in the UserAgents file once at the beginning and caching it, rather than on every request. The current way might slow things down a bunch, particularly if there are a lot of agents in the file.

Reply:
Also.. is s.RandomUserAgent set from the parameters? At a quick skim of the 'changed files' here on GH, I can't see that parameter. Only the prefix/suffix.

Reply:
That's true. Where should I put it then?

I thought in SetupDir(), by adding `UserAgents    []string` to the State and using it to store the array of User-Agents.

But in this case, what do we do with the error? Panic? (as I put in the example), and also, this can't be done there (in SetupDir):

`        s.Printer = PrintDirResult
        s.Processor = ProcessDirEntry
        s.Setup = SetupDir`

As the ProcessDirEntry is run first in `statehelpers.go` and it already requires the User-Agent to be set.

Reply:
I think some of the checks belong in ValidateState() or ValidateDirModeState() if they aren't already: https://github.com/OJ/gobuster/blob/master/libgobuster/statehelpers.go

Adding `UserAgents []string` to the state would be the correct choice I think.

The 'easy' choice is to panic here, the 'better' choice is probably to add support for an 'error' return value to `SetupDir` as well. Which would also require similar to be done for `SetupDns`. And then handled somehow wherever they are called from:
* Added to the definition of `SetupFunc`  here: https://github.com/OJ/gobuster/blob/master/libgobuster/state.go#L24
* The error handling changed from checking a bool to checking for error here. Not actually sure what the bool is representing here.. we may need both it and the error..? But I think it's just meant to be a 'rudimentary' error check: https://github.com/OJ/gobuster/blob/master/libgobuster/state.go#L80
* There is a slight chance there are other changes required, these were the only ones I saw when looking just now though.

Bonus points if you then add an error return to Process: https://github.com/OJ/gobuster/blob/master/libgobuster/state.go#L76

And the associated error check for that at https://github.com/OJ/gobuster/blob/master/main.go#L76

Is `ProcessDirEntry`  actually run in `statehelpers`? The only line I see is https://github.com/OJ/gobuster/blob/master/libgobuster/statehelpers.go#L53 which I believe is just storing the pointer to the function. It's called elsewhere. So I think this would be fine?